### ISSUE TITLE
Export Chat Images to Chatlog

### ISSUE THREAD

**Issue by rkathey** [RPTools/maptool#115]
Currently, the images are not saved as part of the chat log export. It show's a broken image link.

**Comment by Azhrei** [comment]
Hm.  Many of the images are embedded inside the MapTool resources JAR file.  There's no way for a web browser to get to them.  Ditto for token images.

If the image is a link to a web site, that would work in the browser without changes.  I suppose a local file could be converted to a file:// URL (or maybe it wouldn't need converting), but then links should already work so that can't be the issue.

I suppose the smiley and token images could be stored as base64-encoded data in a <script> element at the bottom of the page and then referenced via javascript...  I've never tried anything like that on the browser side but it seems like it could work...

**Comment by rkathey** [comment]
Ideally, the images would be placed in a files folder at the same level as the generated HTML. I assume this would mean extracting the images from the resource library and replacing the links to the MapTool assets with the location in the files folder.

**Comment by Azhrei** [comment]
But every time the chat is saved, do the new images overwrite any already there?  If so, old chat logs may not show the right image any more.  That's why I consider storing smileys (which are very tiny) as base64-encoded images directly inside the HTML -- now the chat log is self-contained.

What about token images?  If the token image is stored separately, how big will it be?  What about newer ones overwriting older ones?  Or maybe the chat log filename is changed to a directory name, something like YYYY-MM-DD-chatlog/index.html and the images are in the directory?  Very saved chat log would get a new directory, but how much space would that consume?  (It would be less space to do it this way than to embed it inside the HTML, but how many chat logs are created over time?  Now we need to implement a self-cleaning facility?  Which we probably ought to have anyway, but still...)

Whatever we decide the final product should be, I'm sure we can manage to produce that product.  It's just a matter of deciding what the final form should look like...

**Comment by JamzTheMan** [comment]
Could also utilize the asset directory but it doesn't lend itself to portability.
What is the goal? To allow chats to be viewed after the fact, locally and remotely?

**Comment by rkathey** [comment]
IIRC when you do a save from a browser, you get <page name>.html and  <page nanme>_files/ directory containing images and other stuff. I would think the chat log export would work the same.

**Comment by rkathey** [comment]
I don't know if that's how I want it to work. It's the way I've seen it work. Another method might be to leverage an exported campaign repository. You'd have to know where the images reside, though. Not sure how much effort that would be.

**Comment by JamzTheMan** [comment]
Well during export, all images except smilies would be in the asset directory and we could link/copy from there. But again, I'd ask, what are you/others doing with the chat logs and how are you using them?

We could provide a chat history viewer in MT. Or do you want to use external tools to view the chat?

**Comment by rkathey** [comment]
I use them as part of my post adventure write-ups. I'd post them online if they looked nice, but they don't.

**Comment by JamzTheMan** [comment]
Mmm, I would say the images by default should link to asset dir to save space but maybe an export command/option to save the images off to a "repository" dir that all the chat saves share. If you do a image dir per chat you are just going to bloat it.
Make it a relative path so you can just upload the images dir and all the chat logs to a server/dropbox

**Comment by rkathey** [comment]
Assuming the user usually picked the same dir, you could create an asset dir there for images. It would simply add images as needed.

**Comment by Azhrei** [comment]
It occurred to me that we could save images by checksum and link to them by checksum.  That way if the image changes, a second chat log could show the new image and the older chat log would still be able to show the old image.

I'd like the HTML to be completely self contained, so if the HTML is distributed the images would go with it automatically.  I'm just not sure that's a good idea since it bloats each HTML image.  The idea of an image directory that all chat logs share makes sense.  But I'd like a user to be able to zip up the directory and distribute to players or post on their site, and copying the entire image directory may be unnecessary if the HTML doesn't refer to all images in the directory.  Maybe the image dir should be per-month or something?  Or maybe it's an option the user provides, the same way the user can change the name of the file used for the chat log?

**Comment by rkathey** [comment]
Well, if I'm following, the image directory would be created with the chatlog. So if the user uses the same directory the all the images would be dumped into a single image directory. If the user picks a different location then a new image directory would be created. So the GM could control the directory size, if required.

**Comment by JamzTheMan** [comment]
I suppose that would work?  Another thought, if you had a single image dir and 6 months of chats, if you open one chat in your browser, then from the browser could you not do a file->save and the browser would re-save the HTML + Images used?  So you can export 1 off chats this way OR just dump the whole image dir to your webserver (or whatever).

**Comment by Azhrei** [comment]
I like that one-off idea, since it wouldn't require any work on our part. :+1:

**Comment by Phergus** [comment]
Though having it done automatically would be nice, it is actually pretty simply to make a saved chat log correctly load images.

Example image line from chat log:
`<img src="asset://0b21e8af0c7e51d2876cef4cd79da722-40">`

Do Search & Replace to switch from asset to file.
Search: `src="asset://"`
Replace: `file://C:/Users/username/.maptool-rptools/assetcache/`

Do Search & Replace to remove the size designation on the end and add a width spec.
Search: `-40">`
Replace: `" width="40">`

**Comment by Azhrei** [comment]
Yeah, just not self-contained.  And even if your example, it would be nice if we could support some kind of `${MAPTOOL_DATADIR}` replacement or something...

Or maybe that could be done in JavaScript?  Prompt the user for the location of their DATADIR, then let JS do the work.  That'd be pretty simple.  Once that's done, then "Save As" should be able to create something that could be moved to other systems.

**Comment by Phergus** [comment]
You want it self contained?   Do a save as from the browser, print into PDF or just open it in OO Write or MS Word.

If you fix it during a save, MT has all the info already and you don't have a prompt a user for the answer to a question they don't know.

**Comment by rkathey** [comment]
Saving as a pdf is acceptable as well.

**Comment by Irarara** [comment]
I only just saw this, but this has been something I've wanted for a long time, and I hacked it in myself recently. Rather than opting to save the assets to a directory to be used by the logs, I base64 encoded the images (scaled to the size they appear at in the log) and inserted them via CSS into the log file. I felt like the ability to sanely share logs individually outweighed the increase in file size (which honestly is not too bad, it's about an extra 5-6KB for a 40x40 token).

I've been hesitating on opening an issue/PR because it feels a little hacky to me (it's mostly lots of text replacement), but it does work and I'd be happy to clean it up and submit a PR if there's any interest. There's also still a few details I'm unsure about like, whether this should supersede the normal log saving or be an additional thing, etc. 

There's also a little optimization that could be made in terms of file size (right now it, it'll include each differently scaled version of an image rather than just the biggest instance, I didn't bother dealing with this at the time because it's not something that comes up for me), and another thought I had was it could potentially be nice to give users the option to set the max size to save the images at themselves, so that for images that could benefit from it, they'd look nicer if the user zoomed in (but that's kind of a weird case, admittedly).

Here's an example (rename .zip to .html): [Image Example.zip](https://github.com/RPTools/maptool/files/5654207/Image.Example.zip)

**Comment by Azhrei** [comment]
In my JavaScript POC, I put the data: URIs in the CSS and referenced them from the HTML, with some JS code to convert the `-##` suffix on asset ids into `width: ## px` CSS styles that I created dynamically.

If the data URIs are included in the CSS to begin with, then the JS code wouldn't be needed (although it could still be included for the purpose of letting the user resize the images, if they wished, but I think this unlikely since many script authors create their output based on the default 40px size).

I think embedding the images isn't a big deal.  If it becomes one, then an option that references the `assetcache` directory might be an alternative, although if the user clears the asset cache, their chat logs revert to not having any images (as Phergus says above, let them open it in a browser and save as PDF if they're that concerned about it).
